src/eradiate/spectral/ckd.py

```diff
@@ -0,0 +1,269 @@
+            on_setattr=None,  # frozen instance: on_setattr must be disabled
+        ),
+        doc='Bin upper spectral bound.\n\nUnit-enabled field (default: ucc["wavelength"]).',
+        type="quantity",
+        init_type="quantity or float",
+    )
+
+    @wmin.validator
+    @wmax.validator
+    def _wbounds_validator(self, attribute, value):
+        if not self.wmin < self.wmax:
+            raise ValueError(
+                f"while validating {attribute.name}: wmin must be lower than wmax"
+            )
+
+    quad: Quad = documented(
+        attrs.field(
+            default=Quad.gauss_legendre(2),
+            repr=lambda x: x.str_summary,
+            validator=attrs.validators.instance_of(Quad),
+        ),
+        doc="Quadrature rule attached to the CKD bin.",
+        type=":class:`.Quad`",
+    )
+
+    bin_set_id: str | None = documented(
+        attrs.field(default=None, converter=str),
+        doc="Id of the bin set used to create this bin.",
+        type="str or None",
+        init_type="str",
+    )
+
+    @property
+    def width(self) -> pint.Quantity:
+        """quantity : Bin spectral width."""
+        return self.wmax - self.wmin
+
+    @property
+    def wcenter(self) -> pint.Quantity:
+        """quantity : Bin central wavelength."""
+        return 0.5 * (self.wmin + self.wmax)
+
+    @property
+    def interval(self) -> P.Interval:
+        """portion.Interval : Closed-open interval corresponding to the bin
+        wavelength interval."""
+        return P.closedopen(self.wmin, self.wmax)
+
+    def pretty_repr(self) -> str:
+        """str : Pretty representation of the bin."""
+        units = ureg.Unit("nm")
+        wrange = (
+            f"[{self.wmin.m_as(units):.1f}, {self.wmax.m_as(units):.1f}] {units:~P}"
+        )
+        quad = self.quad.pretty_repr()
+        return f"{wrange}: {quad}"
+
+    def spectral_indices(self) -> t.Generator[CKDSpectralIndex]:
+        for _, value in enumerate(self.quad.eval_nodes(interval=[0.0, 1.0])):
+            yield CKDSpectralIndex(w=self.wcenter, g=value)
+
+
+# ------------------------------------------------------------------------------
+#                              Bin set data class
+# ------------------------------------------------------------------------------
+
+
+@parse_docs
+@attrs.define(eq=False, frozen=True, slots=True)
+class BinSet:
+    """
+    A data class representing a bin set used in CKD mode.
+    """
+
+    bins: list[Bin] = documented(
+        attrs.field(
+            converter=list,
+            validator=attrs.validators.deep_iterable(
+                member_validator=attrs.validators.instance_of(Bin)
+            ),
+            repr=lambda x: "\n        "
+            + "\n        ".join(
+                [
+                    bin.pretty_repr()
+                    for bin in sorted(x, key=lambda b: b.wcenter.m_as("nm"))
+                ]
+            )
+            + "\n",
```

Comment:
Please avoid multiline reprs, they break easily either the inline or the pretty repr of the object.